pkg/operator/vitesstopo/prune_keyspaces.go

```diff
@@ -0,0 +1,92 @@
+package vitesstopo
+
+import (
+	"context"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"vitess.io/vitess/go/vt/logutil"
+	"vitess.io/vitess/go/vt/topo"
+	"vitess.io/vitess/go/vt/wrangler"
+
+	planetscalev2 "planetscale.dev/vitess-operator/pkg/apis/planetscale/v2"
+	"planetscale.dev/vitess-operator/pkg/operator/results"
+)
+
+type PruneKeyspacesParams struct {
+	// EventObj holds the object type that the recorder will use when writing events.
+	EventObj   runtime.Object
+	TopoServer *topo.Server
+	Recorder   record.EventRecorder
+	// Keyspaces is a current list of KeySpaceTemplates from the cluster spec.
+	Keyspaces []planetscalev2.VitessKeyspaceTemplate
+	// OrphanedKeyspaces is a list of unwanted keyspaces that could not be turned down.
+	OrphanedKeyspaces map[string]*planetscalev2.OrphanStatus
+}
+
+// PruneKeyspaces will prune keyspaces that exist but shouldn't anymore.
+func PruneKeyspaces(ctx context.Context, p PruneKeyspacesParams) (reconcile.Result, error) {
+	resultBuilder := &results.Builder{}
+
+	// Make a map from keyspace name (as Vitess calls them) back to the keyspace spec.
+	desiredKeyspaces := make(map[string]*planetscalev2.VitessKeyspaceTemplate, len(p.Keyspaces))
+	for i := range p.Keyspaces {
+		keyspace := &p.Keyspaces[i]
+		desiredKeyspaces[keyspace.Name] = keyspace
```

Comment:
```suggestion
		desiredKeyspaces[keyspace.Name] = &p.Keyspaces[i]
```

Reply:
Nit: Combine these two lines?

Reply:
good idea - I had simply moved the code but I agree that it's an unnecessary extra line.

Reply:
Done.